core/src/main/java/org/elasticsearch/action/resync/TransportResyncReplicationAction.java

```diff
@@ -0,0 +1,158 @@
+                                           Supplier<ResyncReplicationRequest> replicaRequest, String executor) {
+        transportService.registerRequestHandler(actionName, request, ThreadPool.Names.SAME, new OperationTransportHandler());
+        // we should never reject resync because of thread pool capacity on primary
+        transportService.registerRequestHandler(transportPrimaryAction,
+            () -> new ConcreteShardRequest<>(request),
+            executor, true, true,
+            new PrimaryOperationTransportHandler());
+        transportService.registerRequestHandler(transportReplicaAction,
+            () -> new ConcreteReplicaRequest<>(replicaRequest),
+            executor, true, true,
+            new ReplicaOperationTransportHandler());
+    }
+
+    @Override
+    protected ResyncReplicationResponse newResponseInstance() {
+        return new ResyncReplicationResponse();
+    }
+
+    @Override
+    protected ReplicationOperation.Replicas newReplicasProxy() {
+        // We treat the resync as best-effort for now and don't mark unavailable shard copies as stale.
+        return new ReplicasProxy();
+    }
+
+    @Override
+    protected WritePrimaryResult<ResyncReplicationRequest, ResyncReplicationResponse> shardOperationOnPrimary(
+        ResyncReplicationRequest request, IndexShard primary) throws Exception {
+        final ResyncReplicationRequest replicaRequest = performOnPrimary(request, primary);
+        return new WritePrimaryResult<>(replicaRequest, new ResyncReplicationResponse(), null, null, primary, logger);
+    }
+
+    public static ResyncReplicationRequest performOnPrimary(ResyncReplicationRequest request, IndexShard primary) {
+        return request;
+    }
+
+    @Override
+    protected WriteReplicaResult shardOperationOnReplica(ResyncReplicationRequest request, IndexShard replica) throws Exception {
+        Translog.Location location = performOnReplica(request, replica);
+        return new WriteReplicaResult(request, location, null, replica, logger);
+    }
+
+    public static Translog.Location performOnReplica(ResyncReplicationRequest request, IndexShard replica) throws Exception {
+        Translog.Location location = null;
+        for (Translog.Operation operation : request.getOperations()) {
+            try {
+                final Engine.Result operationResult = replica.applyTranslogOperation(operation, Engine.Operation.Origin.REPLICA,
+                    update -> {
+                        throw new TransportReplicationAction.RetryOnReplicaException(replica.shardId(),
+                            "Mappings are not available on the replica yet, triggered update: " + update);
+                    });
+                location = syncOperationResultOrThrow(operationResult, location);
+            } catch (Exception e) {
+                // if its not a failure to be ignored, let it bubble up
+                if (!TransportActions.isShardNotAvailableException(e)) {
```

Comment:
I copied it over from TransportShardBulkAction to ensure that it is consistent with that one. I think it makes sense to let it bubble up. If we change the behavior for TransportShardBulkAction though, it should be a separate follow-up PR.